### PR TITLE
feat(tools): add image support to read_file tool

### PR DESCRIPTION
**Pull request by samhvw8**
### Related GitHub Issue

Closes: #5169 

### Description
 
 - Add support for reading and displaying image files (PNG, JPG, JPEG, GIF, WebP, SVG, BMP, ICO, TIFF)
- Implement readImageAsDataUrl function to convert images to base64 data URLs with proper MIME types
- Return multi-part responses containing both XML metadata and image data
- Add dimension extraction for PNG files when possible
- Add comprehensive test coverage for image reading functionality including format detection, error handling, and edge cases
- Maintain backward compatibility for non-image binary files

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [ ] **Self-Review**: I have performed a thorough self-review of my code.
- [ ] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [ ] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [ ] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### CODE REVIEW

File: src/core/tools/readFileTool.ts [hunk 1]
```diff
@@ -14,6 +14,58 @@ import { readLines } from "../../integrations/misc/read-lines"
 import { extractTextFromFile, addLineNumbers, getSupportedBinaryFormats } from "../../integrations/misc/extract-text"
 import { parseSourceCodeDefinitionsForFile } from "../../services/tree-sitter"
 import { parseXml } from "../../utils/xml"
+import * as fs from "fs/promises"
+
+/**
+ * Maximum allowed image file size in bytes (5MB)
+ */
+const MAX_IMAGE_FILE_SIZE_BYTES = 5 * 1024 * 1024
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts]: Could we make the maximum image file size configurable? Different use cases might require different limits. Perhaps this could be a setting that users can adjust based on their needs and system capabilities?

File: src/core/tools/__tests__/readFileTool.spec.ts [hunk 2]
```diff
@@ -520,3 +574,325 @@ describe("read_file tool XML output structure", () => {
...
+			// Execute
+			const result = await executeReadImageTool()
+
+			// Verify - should still create valid data URL
+			expect(Array.isArray(result)).toBe(true)
+			const imagePart = (result as any[]).find((p) => p.type === "image")
+			expect(imagePart).toBeDefined()
+			expect(imagePart.source.media_type).toBe("image/png")
+			expect(imagePart.source.data).toBe("")
+		})
```
Comment (daniel-lxs) [inline src/core/tools/__tests__/readFileTool.spec.ts:1667]: What happens if a user tries to read multiple large image files simultaneously? Would it be worth adding a test to ensure the tool handles memory constraints gracefully when processing multiple images at once?

File: src/core/tools/readFileTool.ts [hunk 3]
```diff
@@ -14,6 +14,58 @@ import { readLines } from "../../integrations/misc/read-lines"
...
+	".gif",
+	".webp",
+	".svg",
+	".bmp",
+	".ico",
+	".tiff",
+	".tif",
+	".avif",
+] as const
+
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts]: There's a library for this - the project already uses `pretty-bytes`. Could we use it here instead of the custom `formatFileSize` function for consistency?

**Review by daniel-lxs — COMMENTED** [review]
Hey @samhvw8, I know that different providers have different implementations for image handling, so I'm concerned this centralized approach might break with some providers.

Should we use the provider specific implementations for reading images?

Let me know what you think.

File: src/core/tools/readFileTool.ts (hunk 1 above)
Comment (samhvw8) [reply src/core/tools/readFileTool.ts]: @daniel-lxs agreed, let me add it

File: src/core/tools/__tests__/readFileTool.spec.ts (hunk 2 above)
Comment (samhvw8) [reply src/core/tools/__tests__/readFileTool.spec.ts:1667]: we have max file to read in concurrent, and max image file size setting (i will add this)

File: src/core/tools/readFileTool.ts (hunk 3 above)
Comment (samhvw8) [reply src/core/tools/readFileTool.ts]: ok let me use it

File: src/core/tools/readFileTool.ts [hunk 4]
```diff
@@ -573,20 +699,35 @@ export async function readFileTool(
...
 
+		// Combine all images: feedback images first, then file images
+		const allImages = [...feedbackImages, ...fileImageUrls]
+
 		// Push the result with appropriate formatting
-		if (statusMessage) {
-			const result = formatResponse.toolResult(statusMessage, feedbackImages)
+		if (statusMessage || allImages.length > 0) {
+			// Always use formatResponse.toolResult when we have a status message or images
+			const result = formatResponse.toolResult(
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts:669]: Is it intentional that we're returning image data without checking if the current AI model supports images? I noticed that `maybeRemoveImageBlocks` in `src/api/transform/image-cleaning.ts` checks `apiHandler.getModel().info.supportsImages` before processing images. Should we add a similar check here to prevent sending image data to models that can't process it?

For example, we could check the provider's capability before including images in the response:
```typescript
const provider = await cline.providerRef.deref();
const supportsImages = provider?.apiHandler?.getModel()?.info?.supportsImages ?? false;
const imagesToInclude = supportsImages ? allImages : [];
```

File: src/core/tools/readFileTool.ts [hunk 5]
```diff
@@ -440,14 +493,82 @@ export async function readFileTool(
...
+									}
+								}
+							}
+
+							// Track file read
+							await cline.fileContextTracker.trackFileContext(relPath, "read_tool" as RecordSource)
+
+							// Store image data URL separately - NOT in XML
+							const noticeText = dimensionsInfo
+								? `Image file (${dimensionsInfo}, ${imageSizeInKB} KB)`
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts]: I see that the oversized image error uses i18n properly, but this regular image notice is hardcoded. For consistency, could we move these strings to the translation files as well?

For example:
```typescript
const noticeText = dimensionsInfo
  ? t("tools:readFile.imageWithDimensions", { dimensions: dimensionsInfo, size: imageSizeInKB })
  : t("tools:readFile.imageWithSize", { size: imageSizeInKB });
```

This would ensure all user-facing strings are translatable and maintain consistency across the codebase.

File: src/core/tools/readFileTool.ts [hunk 6]
```diff
@@ -573,20 +699,35 @@ export async function readFileTool(
 			}
 		}
 
+		// Combine all images: feedback images first, then file images
+		const allImages = [...feedbackImages, ...fileImageUrls]
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts:660]: While you mentioned existing settings for max concurrent file reads, I'm curious about the specific handling when multiple large image files are read simultaneously. Could this potentially cause memory issues if someone reads a folder with many images just under the size limit?

Would it be helpful to add:
1. A test case for reading multiple large images
2. Documentation about the memory implications
3. Perhaps a total memory limit across all images in a single read operation?

This could help prevent edge cases where the individual file limit is respected but the total memory usage becomes problematic.

**Review by daniel-lxs — COMMENTED** [review]
Hey @samhvw8 I took another look, I left some suggestions to make this implementation a bit more robust.

File: src/core/tools/readFileTool.ts [hunk 7]
```diff
@@ -573,20 +743,38 @@ export async function readFileTool(
...
+
+		// Use the supportsImages check from the beginning of the function
+		const imagesToInclude = supportsImages ? allImages : []
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts]: I noticed that while we check `supportsImages` at the beginning of the function (line 147), this value could theoretically change if the model is switched during execution. Should we consider moving this check closer to where we actually decide to include images?

The current implementation is likely fine for most cases, but for extra safety, we could store the model info check result and use it consistently throughout the function execution.

File: src/core/tools/readFileTool.ts [hunk 8]
```diff
@@ -14,6 +14,64 @@ import { readLines } from "../../integrations/misc/read-lines"
...
+		".jpg": "image/jpeg",
+		".jpeg": "image/jpeg",
+		".gif": "image/gif",
+		".webp": "image/webp",
+		".svg": "image/svg+xml",
+		".bmp": "image/bmp",
+		".ico": "image/x-icon",
+		".tiff": "image/tiff",
+		".tif": "image/tiff",
+		".avif": "image/avif",
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts]: I noticed that `.avif` is included in `SUPPORTED_IMAGE_FORMATS` (line 44) but there's no corresponding MIME type mapping here. This means AVIF images will default to `image/png` which isn't correct.

Could we add:
```typescript
".avif": "image/avif",
```

to ensure AVIF images are properly identified?

File: src/core/tools/readFileTool.ts [hunk 9]
```diff
@@ -14,6 +14,64 @@ import { readLines } from "../../integrations/misc/read-lines"
...
+/**
+ * Reads an image file and returns both the data URL and buffer
+ */
+async function readImageAsDataUrlWithBuffer(filePath: string): Promise<{ dataUrl: string; buffer: Buffer }> {
+	const fileBuffer = await fs.readFile(filePath)
+	const base64 = fileBuffer.toString("base64")
+	const ext = path.extname(filePath).toLowerCase()
+
+	// Map extensions to MIME types
+	const mimeTypes: Record<string, string> = {
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts]: Consider extracting this MIME type mapping to a module-level constant for better maintainability:

```typescript
const IMAGE_MIME_TYPES: Record<string, string> = {
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
  ".svg": "image/svg+xml",
  ".bmp": "image/bmp",
  ".ico": "image/x-icon",
  ".tiff": "image/tiff",
  ".tif": "image/tiff",
  ".avif": "image/avif", // Don't forget to add this!
}
```

This would make it easier to maintain the mapping between supported formats and their MIME types in one place.

File: src/core/tools/readFileTool.ts [hunk 10]
```diff
@@ -440,14 +507,112 @@ export async function readFileTool(
...
+
+							// Track memory usage for this image
+							totalImageMemoryUsed += imageSizeInMB
+
+							const { dataUrl: imageDataUrl, buffer } = await readImageAsDataUrlWithBuffer(fullPath)
+							const imageSizeInKB = Math.round(imageStats.size / 1024)
+
+							// For images, get dimensions if possible
+							let dimensionsInfo = ""
+							if (fileExtension === ".png") {
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts]: I'm curious about the decision to only extract dimensions for PNG files. Is this intentional?

While PNG dimensions are relatively easy to extract from the IHDR chunk, other formats also have accessible dimension data:
- **JPEG**: Dimensions in SOF0/SOF2 markers
- **GIF**: Dimensions at bytes 6-9 of the header
- **BMP**: Dimensions at specific offsets in the header
- **WebP**: Dimensions in the VP8/VP8L/VP8X chunk

Would it be valuable to extend dimension extraction to other common formats? If not then should we keep the one for PNG files?

File: src/core/tools/readFileTool.ts (hunk 9 above)
Comment (samhvw8) [reply src/core/tools/readFileTool.ts]: ok let me do it

File: webview-ui/src/i18n/locales/en/settings.json [hunk 11]
```diff
@@ -472,6 +472,16 @@
 			"lines": "lines",
 			"always_full_read": "Always read entire file"
 		},
+		"maxImageFileSize": {
+			"label": "Max image file size",
+			"mb": "MB",
+			"description": "Maximum size (in MB) for image files that can be processed by the read file tool."
+		},
+		"maxTotalImageMemory": {
```
Comment (daniel-lxs) [inline webview-ui/src/i18n/locales/en/settings.json]: I feel like this new setting might be hard to understand, and an incorrect value could make Roo Code crash. Should we add a sensible limit ourselves? Maybe we can instead limit the amount of images being read simultaneously?

File: src/core/tools/readFileTool.ts [hunk 12]
```diff
@@ -440,14 +461,101 @@ export async function readFileTool(
...
+								updateFileResult(relPath, {
+									xmlContent: `<file><path>${relPath}</path>\n<notice>${notice}</notice>\n</file>`,
+								})
+								continue
+							}
+
+							// Check if adding this image would exceed total memory limit
+							const imageSizeInMB = imageStats.size / (1024 * 1024)
+							if (totalImageMemoryUsed + imageSizeInMB > maxTotalImageMemory) {
+								const notice = `Image skipped to prevent memory issues. Total image memory would exceed ${maxTotalImageMemory}MB limit (current: ${totalImageMemoryUsed.toFixed(1)}MB, this file: ${imageSizeInMB.toFixed(1)}MB). Consider reading fewer images at once or reducing image sizes.`
```
Comment (daniel-lxs) [inline src/core/tools/readFileTool.ts]: This message is a bit confusing, since the model can't reduce the "memory" of the images and also "memory" is a bit of a vague term. We could change it to something like this:

```suggestion
								const notice = `Image skipped to avoid memory limit (${maxTotalImageMemory}MB). Current: ${totalImageMemoryUsed.toFixed(1)}MB + this file: ${imageSizeInMB.toFixed(1)}MB. Try fewer or smaller images.`;
```

What do you think?

File: webview-ui/src/i18n/locales/en/settings.json (hunk 11 above)
Comment (samhvw8) [reply webview-ui/src/i18n/locales/en/settings.json]: @daniel-lxs the setting will have the same value with concurrent file read setting value

**Comment by daniel-lxs** [comment]
Hey @samhvw8 thank you for addressing the review! I noticed an unit test is failing, can you take a look?

**Comment by samhvw8** [comment]
@daniel-lxs  i fixed the test can you review it again

**Review by daniel-lxs — APPROVED** [review]
Thank you @samhvw8
